### PR TITLE
CloudFormation Registry resource schema support (--registry-schemas)

### PR DESCRIPTION
**Pull request by PatMyron**
https://github.com/aws-cloudformation/cfn-python-lint/issues/1277

[CloudFormation Registry announcement](https://aws.amazon.com/blogs/aws/cloudformation-update-cli-third-party-resource-support-registry/)
[Resource Schema repo](https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/)

---

- [x] jsonschema.validate -> [`RuleMatch`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/docs/getting_started/rules.md#rule-matches) format (**only returns one error per resource + exact context isn't great**)
  - [ ] investigate [iterating all ValidationErrors](https://python-jsonschema.readthedocs.io/en/latest/validate/#jsonschema.IValidator.iter_errors) per resource
- [x] generalize beyond 1 resource type
- [ ] automate updating [PUBLIC resource schemas](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resource-type-schemas.html)
  - [ ] with caching
- [x] handle CloudFormation template syntax
  - [x] https://github.com/aws-cloudformation/cfn-python-lint/blob/338d5fb29a2f47069f0b1091fc420f6dd9e6013d/src/cfnlint/helpers.py#L77
  - [x] https://github.com/aws-cloudformation/cfn-python-lint/blob/338d5fb29a2f47069f0b1091fc420f6dd9e6013d/src/cfnlint/helpers.py#L126
  - [x] https://github.com/aws-cloudformation/cfn-python-lint/blob/338d5fb29a2f47069f0b1091fc420f6dd9e6013d/src/cfnlint/helpers.py#L66
  - [x] https://github.com/aws-cloudformation/cfn-python-lint/blob/1a6657bafb5330ff8e68ae99538ff99f42bcd0bc/src/cfnlint/helpers.py#L65
  - [ ] [JSON schema type checking stricter than CloudFormation's](https://github.com/aws-cloudformation/cfn-python-lint/issues/547)
- [x] allow non-PUBLIC resource schemas to be passed in
  - [x] Update [`E3001`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/docs/rules.md#E3001) to include these resource types:https://github.com/aws-cloudformation/cfn-python-lint/blob/93659fba5cf86ee34177c55b28df0aa9140b7dc9/src/cfnlint/rules/resources/Configuration.py#L105
  - [x] write [`list-types`](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/list-types.html) / [`describe-type`](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/describe-type.html) script to make it easier:
    - [ ] [need to fix CloudFormation paginators](https://github.com/boto/botocore/pull/2177)
```python
import boto3
for page in boto3.client('cloudformation').get_paginator('list_types').paginate():
    for resource_type in page['TypeSummaries']:
        with open(resource_type['TypeName'].replace("::", "-").lower() + '.json', 'w') as f:
            print(boto3.client('cloudformation').describe_type(Arn=resource_type['TypeArn'])['Schema'], file=f)
```

### CODE REVIEW

File: src/cfnlint/rules/resources/ResourceSchema.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+class ResourceSchema(CloudFormationLintRule):
+    id = 'E3000'
+    shortdesc = 'Resource schema'
+    description = 'CloudFormation Registry resource schema validation'
+    source_url = 'https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/'
+    tags = ['resources']
+
+    def match(self, cfn):
+        matches = []
+        for resource_schema in [os.path.basename(f) for f in glob('src/cfnlint/data/ResourceSchemas/*.json')]:
```
Comment (jotompki) [inline src/cfnlint/rules/resources/ResourceSchema.py:23]: so this would assume all the potential schemas are downloaded?

Comment (PatMyron) [reply src/cfnlint/rules/resources/ResourceSchema.py:23]: wrote [`list-types`](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/list-types.html) / [`describe-type`](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/describe-type.html) script to make it easy, but don't think it makes sense to call every runtime due to 
* credentials and networking requirements
* ability to provide additional resource schemas

File: src/cfnlint/rules/resources/ResourceSchema.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+"""
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: MIT-0
+"""
+import os
+import re
+from glob import glob
+from jsonschema import validate, ValidationError
```
Comment (kddejong) [inline src/cfnlint/rules/resources/ResourceSchema.py:8]: One of the big problems we are going to have with using this package is handling `Fn::If` or other intrinsic functions. Either way we are going to have to handle it outside of the schema validation which we have logic to handle or we are going to have to rewrite these validation functions that this package provides to handle these things differently.

**Review by kddejong — COMMENTED** [review]
I'm still not sure the standard jsonschema packages are going to fit the CloudFormation functionality.  We can expand the schema from the registry to support the intrinsic functions when we read them but the possibilities are massive.

File: src/cfnlint/rules/resources/ResourceSchema.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    source_url = 'https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/'
+    tags = ['resources']
+
+    def match(self, cfn):
+        matches = []
+        for resource_schema in [os.path.basename(f) for f in glob('src/cfnlint/data/ResourceSchemas/*.json')]:
+            resource_type = load_resource(ResourceSchemas, resource_schema)['typeName']
+            for resource_name, resource_values in cfn.get_resources([resource_type]).items():
+                properties = resource_values.get('Properties', {})
+                if not re.match(REGEX_DYN_REF, str(properties)) and not any(x in str(properties) for x in PSEUDOPARAMS + UNCONVERTED_SUFFIXES) and FN_PREFIX not in str(properties):
```
Comment (kddejong) [inline src/cfnlint/rules/resources/ResourceSchema.py:27]: Looks like this says ignore intrinsic function errors  but doesn't handle more complex Fn::If scenarios.
```
ResourceName:
   Type: AWS::CloudFormation::StackSet
   Properties:
     Fn::If:
     - ConditionName
     - PropertiesScenarioTrue
     - PropertiesScenarioFalse
```
We basically don't get a failure even though there could be one in the provided scenario.

File: src/cfnlint/rules/resources/ResourceSchema.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+                    try:
+                        validate(properties, load_resource(ResourceSchemas, resource_schema))
+                    except ValidationError as e:
+                        matches.append(RuleMatch(['Resources', resource_name], e.message))
```
Comment (kddejong) [inline src/cfnlint/rules/resources/ResourceSchema.py:31]: Should add the Path of `Properties` too but don't like making it so generic.

Comment (PatMyron) [reply src/cfnlint/rules/resources/ResourceSchema.py:31]: > Should add the Path of `Properties` too

was thinking not every resource would have `Properties`, but [`RuleMatch`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/docs/getting_started/rules.md#rule-matches) paths seem to roll up to the next levels anyway?

File: src/cfnlint/rules/resources/ResourceSchema.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    source_url = 'https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/'
+    tags = ['resources']
+
+    def match(self, cfn):
+        matches = []
+        for schema in REGISTRY_SCHEMAS:
+            resource_type = schema['typeName']
+            for resource_name, resource_values in cfn.get_resources([resource_type]).items():
+                properties = resource_values.get('Properties', {})
+                if not re.match(REGEX_DYN_REF, str(properties)) and not any(x in str(properties) for x in PSEUDOPARAMS + UNCONVERTED_SUFFIXES) and FN_PREFIX not in str(properties):
```
Comment (miparnisari) [inline src/cfnlint/rules/resources/ResourceSchema.py:24]: line 24 is hard to read for me. what is `REGEX_DYN_REF`? 

`not any(x in str(properties) for x in PSEUDOPARAMS + UNCONVERTED_SUFFIXES) and FN_PREFIX not in str(properties)` 

can this be assigned to a variable `no_intrinsic_function_errors`?

Comment (PatMyron) [reply src/cfnlint/rules/resources/ResourceSchema.py:25]: > what is `REGEX_DYN_REF`?

[dynamic references](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references.html):
https://github.com/aws-cloudformation/cfn-python-lint/blob/f90768da4513a5ee8d8782d86bc2256419f5f209/src/cfnlint/helpers.py#L77-L79

> can this be assigned to a variable `no_intrinsic_function_errors`?

added a comment above that line:
https://github.com/aws-cloudformation/cfn-python-lint/blob/82a5381b9e6ee375af164e8ae1dc1d74f183d63d/src/cfnlint/rules/resources/ResourceSchema.py#L24

File: README.md [hunk 6]
```diff
@@ -166,6 +166,7 @@ Optional parameters:
 | -u, --update-specs | | | Update the [CloudFormation Resource Specifications](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-resource-specification.html).  You may need sudo to run this.  You will need internet access when running this command |
 | -o, --override-spec | | filename | Spec-style file containing custom definitions. Can be used to override CloudFormation specifications. More info [here](#customize-specifications) |
 | -g, --build-graph | |  | Creates a file in the same directory as the template that models the template's resources in [DOT format](https://en.wikipedia.org/wiki/DOT_(graph_description_language)) |
+| -s, --registry-schemas | | | one or more directories of [CloudFormation Registry](https://aws.amazon.com/blogs/aws/cloudformation-update-cli-third-party-resource-support-registry/) [Resource Schemas](https://github.com/aws-cloudformation/aws-cloudformation-resource-schema/)
```
Comment (kddejong) [inline README.md:169]: `Downloads 3rd party registry resource type schemas into one or more directories.  Requires AWS Credentials` ?

File: src/cfnlint/core.py [hunk 7]
```diff
@@ -39,7 +40,7 @@ class UnexpectedRuleException(CfnLintExitException):
     """When processing a rule fails in an unexpected way"""
 
 
-def run_cli(filename, template, rules, regions, override_spec, build_graph, mandatory_rules=None):
+def run_cli(filename, template, rules, regions, override_spec, build_graph, registry_schemas, mandatory_rules=None):
```
Comment (kddejong) [inline src/cfnlint/core.py:43]: outside packages shouldn't be using run_cli by any concerns with the ordering here?  I've been adding parameters and setting a default value to keep some backwards compatibility.

**Review by kddejong — CHANGES_REQUESTED** [review]
Few changes but looks good to me

File: src/cfnlint/core.py (hunk 7 above)
Comment (PatMyron) [reply src/cfnlint/core.py:43]: only found this from searching: https://github.com/Cloudzero/samwise/blob/793463e0774f9d1c702f4252377f19fa3408a558/samwise/__main__.py#L129 which looks already broken from https://github.com/aws-cloudformation/cfn-python-lint/pull/1411/
cc @silvexis

Comment (PatMyron) [reply src/cfnlint/core.py:43]: have been generally contemplating compatibility for reliance on [undocumented internal functions/variables](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/docs/getting_started/integration.md)

**Comment by PatMyron** [comment]
Example messages:
```bash
[cfn-lint] E3000 'RequiredEnumProperty' is a required property
[cfn-lint] E3000 Additional properties are not allowed ('InvalidProperty' was unexpected)
[cfn-lint] E3000 'string' is not of type 'integer'
[cfn-lint] E3000 200 is greater than the maximum of 100
[cfn-lint] E3000 -1 is less than the minimum of 0
[cfn-lint] E3000 'd' is not one of ['a', 'b', 'c']
[cfn-lint] E3000 'name' does not match 'arn:.*'
```
Screen Shot 2020-12-04 at 2 54 46 PM
<details>
  <summary>schema/template</summary>

```json
{
  "typeName" : "Third::Party::Type",
  "description" : "",
  "additionalProperties" : false,
  "properties" : {
    "RequiredEnumProperty" : {
      "type" : "string",
      "enum": [
        "a",
        "b",
        "c"
      ]
    },
    "PatternProperty" : {
      "type" : "string",
      "pattern": "arn:.*"
    },
    "RangeProperty" : {
      "type": "integer",
      "minimum": 0,
      "maximum": 100
    }
  },
  "required" : ["RequiredEnumProperty"],
  "primaryIdentifier" : [ "/properties/RequiredEnumProperty" ]
}
```
```yaml
Resources:
  ResourceMissingRequiredProperty:
    Type: Third::Party::Type
  ResourceMissingRequiredProperty2:
    Type: Third::Party::Type
    Properties:
      RangeProperty: 50
  ResourceWithInvalidProperty:
    Type: Third::Party::Type
    Properties:
      RequiredEnumProperty: required
      RangeProperty: 50
      InvalidProperty: invalid
  ResourceWithWrongPrimitiveType:
    Type: Third::Party::Type
    Properties:
      RequiredEnumProperty: a
      RangeProperty: string
  ResourceWithValueOutsideMinMaxRange:
    Type: Third::Party::Type
    Properties:
      RequiredEnumProperty: a
      RangeProperty: 200
  ResourceWithValueOutsideMinMaxRange2:
    Type: Third::Party::Type
    Properties:
      RequiredEnumProperty: a
      RangeProperty: -1
  ResourceWithInvalidValue:
    Type: Third::Party::Type
    Properties:
      RequiredEnumProperty: d
      RangeProperty: 50
  ResourceWithInvalidPattern:
    Type: Third::Party::Type
    Properties:
      RequiredEnumProperty: a
      PatternProperty: name
  ValidResource:
    Type: Third::Party::Type
    Properties:
      RequiredEnumProperty: a
      RangeProperty: 50
  ResourceWithUnresolvableSyntax:
    Type: Third::Party::Type
    Properties:
      RequiredEnumProperty: !Ref ValidResource
      RangeProperty: '{{resolve:secretsmanager:secret:secret:secret}}'
```

</details>

**Comment by jasonmimick** [comment]
Need to add this new parameter to the config file json schema:
`cfnlint/data/CfnLintCli/config/schema.json`
needs
```vim
...
 "ignore_templates": {
  79       "description": "Templates to ignore",
  80       "items": {
  81         "type": "string"
  82       },
  83       "type": "array"
  84     },
  85     "registry_schemas": {
  86       "description": "one or more directories of CloudFormation Registry Resource Schemas",
  87       "items": {
  88         "type": "string"
  89       },
  90       "type": "array"
  91     }
```
```

**Pull request merged**